workflow/util/util.go

```diff
@@ -799,50 +788,426 @@ func deletePodNodeDuringRetryWorkflow(wf *wfv1.Workflow, node wfv1.NodeStatus, d
+		}
+
+		if foundNode, shouldReset := should(curr); foundNode {
+			if shouldReset {
+				resetFunc(curr.n.ID, true)
+			}
+			return curr, nil
+		}
+		curr = curr.parent
+	}
+}
+
+func getTillBoundaryFn(boundaryID string) tillFn {
+	return func(n *node) (bool, bool) {
+		return n.n.ID == boundaryID, n.n.BoundaryID != ""
+	}
 }
 
-func resetConnectedParentGroupNodes(oldWF *wfv1.Workflow, newWF *wfv1.Workflow, currentNode wfv1.NodeStatus, resetParentGroupNodes []string) (*wfv1.Workflow, []string) {
-	currentNodeID := currentNode.ID
+func consumeBoundaries(n *node, resetFunc resetFn) (*node, error) {
+	curr := n
 	for {
-		currentNode, err := oldWF.Status.Nodes.Get(currentNodeID)
+		if curr == nil {
+			return curr, nil
+		}
+		if curr.parent != nil && curr.parent.n.Type == wfv1.NodeTypeStepGroup {
+			resetFunc(curr.parent.n.ID, true)
+		}
+		seekingBoundaryID := curr.n.BoundaryID
+		if seekingBoundaryID == "" {
+			return curr.parent, nil
+		}
+		var err error
+		curr, err = consumeTill(curr, getTillBoundaryFn(seekingBoundaryID), resetFunc)
 		if err != nil {
-			log.Panicf("dying due to inability to obtain node for %s, panicking", currentNodeID)
+			return nil, err
+		}
+	}
+}
+
+func consumeStepGroup(n *node, resetFunc resetFn) (*node, error) {
+	return consumeTill(n, getTillFnNodeType(wfv1.NodeTypeStepGroup), resetFunc)
+}
+
+func consumeSteps(n *node, resetFunc resetFn) (*node, error) {
+	n, err := consumeTill(n, getTillFnNodeType(wfv1.NodeTypeSteps), resetFunc)
+	if err != nil {
+		return nil, err
+	}
+	return consumeBoundaries(n, resetFunc)
+}
+
+func consumeTaskGroup(n *node, resetFunc resetFn) (*node, error) {
+	return consumeTill(n, getTillFnNodeType(wfv1.NodeTypeTaskGroup), resetFunc)
+}
+
+func consumeDAG(n *node, resetFunc resetFn) (*node, error) {
+	n, err := consumeTill(n, getTillFnNodeType(wfv1.NodeTypeDAG), resetFunc)
+	if err != nil {
+		return nil, err
+	}
+	return consumeBoundaries(n, resetFunc)
+}
+
+func consumePod(n *node, resetFunc resetFn, addToDelete deleteFn) (*node, error) {
+	// this sets to reset but resets are overridden by deletes in the final FormulateRetryWorkflow logic.
+	curr, err := consumeTill(n, getTillFnNodeType(wfv1.NodeTypePod), resetFunc)
+	if err != nil {
+		return nil, err
+	}
+	addToDelete(curr.n.ID, true)
+	children := getChildren(curr)
+	for childID := range children {
+		addToDelete(childID, true)
+	}
+	return curr, nil
+}
+
+func resetPath(allNodes []*node, toNode string) (map[string]bool, map[string]bool, error) {
+	nodes, err := singularPath(allNodes, toNode)
+
+	curr := nodes[len(nodes)-1]
+	if len(nodes) > 0 {
+		// remove toNode
+		nodes = nodes[:len(nodes)-1]
+	}
+
+	nodesToDelete := getChildren(curr)
+	nodesToDelete[curr.n.ID] = true
+
+	nodesToReset := make(map[string]bool)
+
+	if err != nil {
+		return nil, nil, err
+	}
+	l := len(nodes)
+	if l <= 0 {
+		return nodesToReset, nodesToDelete, nil
+	}
+
+	addToReset := func(nodeID string, addToNode bool) {
```

Comment:
`addToNode` is always `false`, so simplify

Reply:
It is actually safe to get rid  of the `addToNode` arg here, the reason being the deletion overrides the resetting. 
Maybe a comment  explaining so and removing this arg is best.